File: integration-tests/validate-vectors.py
#!/usr/bin/env python3
"""
Cross-collector HMAC validation — Python side.

Validates that the Python collector's sign matches the spec test
vectors exactly.

New format: "{timestamp}\\n{method}\\n{path}\\n{body}"
Signature is raw hex (no sha256= prefix). No canonicalization.

Run: python3 integration-tests/validate-vectors.py
"""

import json
import os
import sys

# Add Python collector to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'collectors', 'python', 'src'))

from ailabsaudit_tracker.tracker import sign, signature_header  # noqa: E402

vectors_path = os.path.join(os.path.dirname(__file__), '..', 'spec', 'test-vectors.json')
with open(vectors_path) as f:
    data = json.load(f)

secret = data['secret']
passed = 0
failed = 0

for v in data['vectors']:
    sig = sign(v['timestamp'], v['method'], v['path'], v['body'], secret)
    header = signature_header(v['timestamp'], v['method'], v['path'], v['body'], secret)
    errors = []

    if sig != v['signature']:
        errors.append(f"  signature mismatch\n    expected: {v['signature']}\n    actual:   {sig}")
    if header != v['signature']:
        errors.append(f"  header mismatch\n    expected: {v['signature']}\n    actual:   {header}")

    if errors:
        failed += 1
        print(f"  FAIL: {v['name']}")
        for e in errors:
            print(e)
    else:
        passed += 1
        print(f"  PASS: {v['name']}")

total = len(data['vectors'])
suffix = f", {failed} FAILED" if failed > 0 else ""
print(f"\nResults: {passed}/{total} passed{suffix}")
sys.exit(1 if failed > 0 else 0)

File: integration-tests/validate-vectors.js
'use strict';

/**
 * Cross-collector HMAC validation script.
 *
 * Validates that the HMAC signing logic matches the spec test vectors.
 * Uses Node.js crypto to replicate the signing format used by all collectors.
 *
 * Format: "{timestamp}\n{method}\n{path}\n{body}"
 * Signature is raw hex (no sha256= prefix). No canonicalization.
 *
 * Run: node integration-tests/validate-vectors.js
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Load test vectors.
const vectorsPath = path.resolve(__dirname, '..', 'spec', 'test-vectors.json');
const data = JSON.parse(fs.readFileSync(vectorsPath, 'utf8'));
const secret = data.secret;

/**
 * HMAC-SHA256 sign — same logic as all collectors.
 */
function sign(timestamp, method, pathStr, body, secret) {
  const stringToSign = `${timestamp}\n${method}\n${pathStr}\n${body}`;
  return crypto.createHmac('sha256', secret).update(stringToSign, 'utf8').digest('hex');
}

let passed = 0;
let failed = 0;

for (const v of data.vectors) {
  const sig = sign(v.timestamp, v.method, v.path, v.body, secret);
  const errors = [];

  if (sig !== v.signature) {
    errors.push(`  signature mismatch\n    expected: ${v.signature}\n    actual:   ${sig}`);
  }

  if (errors.length > 0) {
    failed++;
    console.log(`  FAIL: ${v.name}`);
    errors.forEach((e) => console.log(e));
  } else {
    passed++;
    console.log(`  PASS: ${v.name}`);
  }
}

const total = data.vectors.length;
console.log(`\nResults: ${passed}/${total} passed${failed > 0 ? `, ${failed} FAILED` : ''}`);
process.exit(failed > 0 ? 1 : 0);
